src/main/java/com/minecolonies/coremod/commands/generalcommands/ScanCommand.java

```diff
@@ -52,41 +51,47 @@ public ScanCommand(@NotNull final String... parents)
     public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenu actionMenu) throws CommandException
     {
         // Will throw ClassCastException if null, but should never be null as these values are required.
+        final EntityPlayerMP player = actionMenu.getPlayerForArgument("player");
         final int x1 = actionMenu.getIntegerForArgument("x1");
         final int y1 = actionMenu.getIntegerForArgument("y1");
         final int z1 = actionMenu.getIntegerForArgument("z1");
         final int x2 = actionMenu.getIntegerForArgument("x2");
         final int y2 = actionMenu.getIntegerForArgument("y2");
         final int z2 = actionMenu.getIntegerForArgument("z2");
+
         final BlockPos from = new BlockPos(x1, y1, z1);
         final BlockPos to = new BlockPos(x2, y2, z2);
-        executeShared(server, sender, from, to);
+        executeShared(server, sender, from, to, player);
     }
 
     @Override
     public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
     {
-//        executeShared(server, sender, from, to);
+        //executeShared(server, sender, from, to);
     }
 
     private void executeShared(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender,
-            @NotNull final BlockPos from, @NotNull final BlockPos to) throws CommandException
+            @NotNull final BlockPos from, @NotNull final BlockPos to, @Nullable final EntityPlayerMP playerArgument) throws CommandException
     {
 
         if (isPlayerOpped(sender))
         {
             server.addScheduledTask(() ->
             {
                 @Nullable final World world = server.getEntityWorld();
-                @NotNull final EntityPlayer player;
-                if (sender instanceof EntityPlayer)
+                @NotNull final EntityPlayerMP player;
+                if(playerArgument != null)
+                {
+                    player = playerArgument;
+                }
+                else if (sender instanceof EntityPlayer)
                 {
-                    player = (EntityPlayer) sender;
+                    player = (EntityPlayerMP) sender;
                 }
                 else
                 {
-                    // Not sure this is allowed in saveStructure()
-                    player = null;
+                    sender.sendMessage(new TextComponentString(SCAN_FAILURE_MESSAGE));
```

Comment:
You should use an error message saying that sender must be a player or player must be provided rather than saying that the scan failed.

There are a lot of commands which will try to calculate the player from the sender.   I wonder if we should centralize this code somehow.   It's the same situation for the commands which try to compute the colony from the sender.  Not sure what the best way to do this would be, though.  I guess at minimum a method like getPlayerFromSender(sender) or getColonyFromSender(sender) in some utility class.

Reply:
Yeah, we should centralize a lot of this code.
Especially the "getColony" part which might get it from sender, playername or colonyId.

Reply:
The colony argument already works with ether "-1" for using the sender, or a colony id for using the the id.

But maybe what we need is a "if not specified, default to..." parameter for arguments.   We could have it be processed in the main execute method after we determine that an argument isn't set. and call a method to set it.   That way it's specified in the command definition rather than java code for each command.

  
 have it call a method and be processed in the main execute method

Reply:
For things taking one or another argument (colony or player), I've been thinking that a Is.REQUIRED_ONLY_ONE_OF("group") option in addition to REQUIRED and OPTIONAL.   That way, the parsing code would handle making sure that the arguments are provided rather than the individual methods.

Reply:
That sounds like a great solution which would avoid a lot of duplicated code already